hooks/tk-multi-publish2/basic/publish_session.py

```diff
@@ -0,0 +1,712 @@
+        :param dict settings: Configured publish settings
+
+        :return: A dictionary of the form::
+
+            {
+                "publish_path": "/path/to/file/to/publish/filename.v0001.ma",
+                "publish_name": "filename.ma",
+                "publish_version": 1,
+                "publish_type": "Maya Scene"
+            }
+        """
+
+        publisher = self.parent
+
+        # by default, extract the version number from the work file for
+        # publishing. use 1 if no version in path
+        version_number = publisher.util.get_version_number(path) or 1
+
+        # publish in place by default
+        publish_path = path
+
+        # the configured publish type
+        publish_type = settings["Publish Type"].value
+
+        # ---- Check to see if templates are configured
+
+        # NOTE: we're using `get_template_by_name` here since the settings are
+        # not configured at the app level.
+
+        work_template = self._get_template("Work file Template", settings)
+        publish_template = self._get_template("Publish file Template", settings)
+
+        if work_template and publish_template:
+
+            self.logger.debug("Work and publish templates are defined")
+
+            # templates are defined, see if the session path matches the work
+            # file template
+            fields = work_template.validate_and_get_fields(path)
+
+            self.logger.debug("Fields extracted from work file: %s" % (fields,))
+
+            if fields:
+
+                # scene path matches the work file template. execute the
+                # "classic" toolkit behavior of constructing the output publish
+                # path and copying the work file to that location
+                fields["TankType"] = publish_type
+
+                # construct the publish path
+                publish_path = publish_template.apply_fields(fields)
+
+                self.logger.debug("Publish path: %s" % (publish_path,))
+
+                # if version number is one of the fields, use it to populate the
+                # publish information, else fall back to the default version,
+                # extracted from the work file above
+                version_number = fields.get("version", version_number)
+
+                self.logger.debug("Version number: %s" % (version_number,))
+
+        # get the publish name for the publish file. this will ensure we get a
+        # consistent name across version publishes of this file.
+        publish_name = publisher.util.get_publish_name(publish_path)
+
+        return {
+            "publish_path": publish_path,
+            "publish_name": publish_name,
+            "publish_version": version_number,
+            "publish_type": publish_type,
+        }
+
+    def _get_template(self, template_name, settings):
+        """Return the configured template for the supplied name.
+
+        TODO: should publish2 populate "settings" with the configured template?
```

Comment:
again, should publish2 handle identifying/validating the configured settings?

Reply:
if you have defined a setting like this:

```
"Publish file Template": {
                 "type": "template",
                 "default": None,
                 "description": "Template path for published work files. Should"
                                "correspond to a template defined in "
                                "templates.yml."
             }
```

e.g. of type `template` i would expect the publisher to resolve it into a template object, such that `settings["Publish file Template"]` returns a template object.

Reply:
ok. currently it does not. Let's discuss a path forward.